File: frontend/src/api/knowledge.ts
import request from '@/request'

export interface KnowledgeDocumentVO {
  id: number
  title: string
  fileName: string
  fileType: string
  fileSize: number
  status: string
  chunkCount: number
  errorMessage?: string
  createdBy: number
  createTime: string
  updateTime: string
}

export interface KnowledgeStatsVO {
  total: number
  readyCount: number
  processingCount: number
  failedCount: number
}

export interface KnowledgeChunkVO {
  chunkIndex: number
  content: string
  title: string
}

export async function uploadKnowledgeDocument(file: File, title?: string) {
  const formData = new FormData()
  formData.append('file', file)
  if (title?.trim()) {
    formData.append('title', title.trim())
  }
  return request.post<{ code: number; data?: KnowledgeDocumentVO; message?: string }>(
    '/knowledge/upload',
    formData,
    {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120_000,
    },
  )
}

export async function listKnowledgeDocuments(
  current = 1,
  pageSize = 10,
  options?: { status?: string; title?: string },
) {
  return request.post<{
    code: number
    data?: {
      records: KnowledgeDocumentVO[]
      total: number
      current: number
      size: number
    }
    message?: string
  }>('/knowledge/list/page', {
    current,
    pageSize,
    ...(options?.status ? { status: options.status } : {}),
    ...(options?.title ? { title: options.title } : {}),
  })
}

export async function getKnowledgeStats() {
  return request.get<{ code: number; data?: KnowledgeStatsVO; message?: string }>(
    '/knowledge/stats',
  )
}

export async function deleteKnowledgeDocument(id: number) {
  return request.post<{ code: number; data?: boolean; message?: string }>(
    '/knowledge/delete',
    { id },
  )
}

export async function reindexKnowledgeDocument(id: number) {
  return request.post<{ code: number; data?: KnowledgeDocumentVO; message?: string }>(
    `/knowledge/reindex/${id}`,
  )
}

export async function updateKnowledgeDocumentTitle(id: number, title: string) {
  return request.post<{ code: number; data?: KnowledgeDocumentVO; message?: string }>(
    '/knowledge/update/title',
    { id, title },
  )
}

export async function searchKnowledgeDocumentsByStatus(status: string) {
  return request.post<{ code: number; data?: KnowledgeDocumentVO[]; message?: string }>(
    '/knowledge/search/status',
    { status },
  )
}

export async function searchKnowledgeDocumentsByTitle(title: string) {
  return request.post<{ code: number; data?: KnowledgeDocumentVO[]; message?: string }>(
    '/knowledge/search/title',
    { title },
  )
}

export async function listKnowledgeDocumentChunks(documentId: number) {
  return request.get<{ code: number; data?: KnowledgeChunkVO[]; message?: string }>(
    `/knowledge/${documentId}/chunks`,
  )
}

export async function batchDeleteKnowledgeDocuments(ids: number[]) {
  return request.post<{ code: number; data?: boolean; message?: string }>(
    '/knowledge/batch/delete',
    { ids },
  )
}

File: python-backend/app/routers/knowledge.py
"""系统知识库路由（管理员）"""

from typing import List, Optional

from databases import Database
from fastapi import APIRouter, Depends, File, Form, UploadFile

from app.database import get_db
from app.deps import require_admin
from app.exceptions import ErrorCode, throw_if
from app.schemas.common import BaseResponse, DeleteRequest
from app.schemas.knowledge import (
    KnowledgeChunkVO,
    KnowledgeDocumentVO,
    KnowledgeQueryRequest,
    KnowledgeUpdateTitleRequest,
    KnowledgeSearchByStatusRequest,
    KnowledgeSearchByTitleRequest,
    KnowledgeStatsVO,
)
from app.schemas.user import LoginUserVO
from app.services.knowledge_service import KnowledgeService

router = APIRouter(prefix="/knowledge", tags=["系统知识库"])


@router.post("/upload", response_model=BaseResponse[KnowledgeDocumentVO])
async def upload_knowledge_document(
    file: UploadFile = File(...),
    title: Optional[str] = Form(None),
    db: Database = Depends(get_db),
    admin: LoginUserVO = Depends(require_admin),
):
    """上传并索引知识库文档（.txt / .md / .docx）"""
    throw_if(not file.filename, ErrorCode.PARAMS_ERROR, "文件名不能为空")
    file_bytes = await file.read()
    service = KnowledgeService(db)
    document = await service.create_and_ingest(
        title=(title or "").strip() or file.filename,
        file_name=file.filename,
        file_bytes=file_bytes,
        admin_user_id=admin.id,
    )
    return BaseResponse.success(data=document, message="上传并索引成功")


@router.post("/list/page", response_model=BaseResponse[dict])
async def list_knowledge_documents(
    request: KnowledgeQueryRequest,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    service = KnowledgeService(db)
    records, total = await service.list_by_page(request)
    return BaseResponse.success(
        data={
            "records": records,
            "total": total,
            "current": request.current,
            "size": request.page_size,
        }
    )


@router.get("/stats", response_model=BaseResponse[KnowledgeStatsVO])
async def get_knowledge_stats(
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    """知识库统计概览（各状态文档数量）"""
    service = KnowledgeService(db)
    stats = await service.get_stats()
    return BaseResponse.success(data=stats)


@router.post("/search/title", response_model=BaseResponse[List[KnowledgeDocumentVO]])
async def search_knowledge_documents_by_title(
    request: KnowledgeSearchByTitleRequest,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    """根据标题关键词搜索知识库文档（不分页）"""
    service = KnowledgeService(db)
    documents = await service.search_by_title(request)
    return BaseResponse.success(data=documents, message="搜索成功")


@router.post("/search/status", response_model=BaseResponse[List[KnowledgeDocumentVO]])
async def search_knowledge_documents_by_status(
    request: KnowledgeSearchByStatusRequest,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    """根据状态搜索知识库文档"""
    service = KnowledgeService(db)
    documents = await service.search_by_status(request)
    return BaseResponse.success(data=documents, message="搜索成功")


@router.get("/{document_id}/chunks", response_model=BaseResponse[List[KnowledgeChunkVO]])
async def list_knowledge_document_chunks(
    document_id: int,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    """查看文档分块内容（向量库中的切片预览）"""
    service = KnowledgeService(db)
    chunks = await service.list_chunks(document_id)
    return BaseResponse.success(data=chunks)


@router.get("/{document_id}", response_model=BaseResponse[KnowledgeDocumentVO])
async def get_knowledge_document(
    document_id: int,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    service = KnowledgeService(db)
    document = await service.get_by_id(document_id)
    throw_if(not document, ErrorCode.NOT_FOUND_ERROR, "文档不存在")
    return BaseResponse.success(data=document)


@router.post("/update/title", response_model=BaseResponse[KnowledgeDocumentVO])
async def update_knowledge_document_title(
    request: KnowledgeUpdateTitleRequest,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    """重命名知识库文档标题"""
    service = KnowledgeService(db)
    document = await service.update_title(request.id, request.title)
    return BaseResponse.success(data=document, message="标题已更新")


@router.post("/delete", response_model=BaseResponse[bool])
async def delete_knowledge_document(
    request: DeleteRequest,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    service = KnowledgeService(db)
    result = await service.delete_document(request.id)
    return BaseResponse.success(data=result, message="删除成功")


@router.post("/reindex/{document_id}", response_model=BaseResponse[KnowledgeDocumentVO])
async def reindex_knowledge_document(
    document_id: int,
    db: Database = Depends(get_db),
    _: LoginUserVO = Depends(require_admin),
):
    service = KnowledgeService(db)
    document = await service.reindex_document(document_id)
    return BaseResponse.success(data=document, message="重建索引成功")
